hephaistus/main-theme/distribution-system/chord/playground/nullist0/lib/src/main/kotlin/org/lia/dht/nullist0/chord/model/ChordScope.kt

```diff
@@ -0,0 +1,37 @@
+package org.lia.dht.nullist0.chord.model
+
+/**
+ * An interface to build identifier of Chord Ring.
+ */
+abstract class ChordScope<Id, Value> {
+    abstract fun AbstractChordNode<Id, Value>.toId(): Id
+
+    abstract fun bitSize(): Int
+
+    abstract fun Int.toId(): Id
+
+    /**
+     * The [Id] in a chord ring, is an element of integers modulo n, where a natural number n is a power of 2.
+     * There is a natural addition in such set, so this function will implement that addition.
+     */
+    abstract infix operator fun Id.plus(other: Id): Id
+
+    /**
+     * The [Id] in a chord ring, is an element of integers modulo n, where a natural number n is a power of 2.
+     * Since there is no order in the such set, this function will implement an order on the integer set,
+     * not on the modulo set. In other words, this will return the natural order of integers.
+     */
+    protected abstract infix operator fun Id.compareTo(other: Id): Int
+
+    infix fun Id.inOpenInterval(pair: Pair<Id, Id>): Boolean  {
+        val id = this@inOpenInterval
```

Comment:
What is meaning of it? 🙂

Reply:
Is this comment about `@inOpenInterval`? This is a language feature of Kotlin. If there is an anonymous `this`, this can be denoted by `@`. In this code, there are two meaning of `this`. One is `this` of `ChordScope` and another is `this` of `Id` because `inOpenInterval` is a function with receiver `Id`. `@inOpenInterval` will denote that `this` is `Id`.

Reply:
Thank you, interesting.

Reply:
For more details, this link will help.

https://kotlinlang.org/docs/this-expressions.html#qualified-this